src/Voron/StorageEnvironment.cs

```diff
@@ -38,8 +38,49 @@ namespace Voron
 {
     public delegate bool UpgraderDelegate(SchemaUpgradeTransactions transactions, int currentVersion, out int versionAfterUpgrade);
 
-    public class StorageEnvironment : IDisposable
+
+    public class StorageEnvironmentSynchronization
     {
+        public struct SynchronizationSection : IDisposable
+        {
+            private readonly Semaphore _context;
+
+            public SynchronizationSection(Semaphore context)
+            {
+                this._context = context;
+                if (this._context != null)
```

Comment:
Under what scenario this can be null and we want to accept that?

Reply:
`None` is the scenario. We don't need to pay the synchronization cost if it is not used.